planner/core/logical_plan_builder.go

```diff
@@ -3784,24 +3786,61 @@ func (ds *DataSource) newExtraHandleSchemaCol() *expression.Column {
 // addExtraPIDColumn add an extra PID column for partition table.
 // 'select ... for update' on a partition table need to know the partition ID
 // to construct the lock key, so this column is added to the chunk row.
-func (ds *DataSource) addExtraPIDColumn(info *extraPIDInfo) {
+func (ds *DataSource) addExtraPIDColumn() {
 	pidCol := &expression.Column{
 		RetType:  types.NewFieldType(mysql.TypeLonglong),
 		UniqueID: ds.ctx.GetSessionVars().AllocPlanColumnID(),
-		ID:       model.ExtraPidColID,
-		OrigName: fmt.Sprintf("%v.%v.%v", ds.DBName, ds.tableInfo.Name, model.ExtraPartitionIdName),
+		ID:       model.ExtraPhysTblID,
+		OrigName: fmt.Sprintf("%v.%v.%v", ds.DBName, ds.tableInfo.Name, model.ExtraPhysTblIdName),
 	}
 
-	ds.Columns = append(ds.Columns, model.NewExtraPartitionIDColInfo())
+	ds.Columns = append(ds.Columns, model.NewExtraPhysTblIDColInfo())
 	schema := ds.Schema()
 	schema.Append(pidCol)
 	ds.names = append(ds.names, &types.FieldName{
 		DBName:      ds.DBName,
 		TblName:     ds.TableInfo().Name,
-		ColName:     model.ExtraPartitionIdName,
-		OrigColName: model.ExtraPartitionIdName,
+		ColName:     model.ExtraPhysTblIdName,
+		OrigColName: model.ExtraPhysTblIdName,
 	})
 	ds.TblCols = append(ds.TblCols, pidCol)
+}
+
+// addExtraPIDColumnWithInfo add an extra PID column for partition table.
+// 'select ... for update' on a partition table need to know the partition ID
+func (ds *DataSource) addExtraPIDColumnWithInfo(info *extraPIDInfo) {
```

Comment:
Can we merge `addExtraPIDColumnWithInfo` and `addExtraPIDColumn` into one? They seem similar.

Reply:
This is still a hack from my side, I hope to get rid of setting any PIDColumn within `buildSelectLock()` , I hope it could only be set when building the logical plan's data source in `buildDataSource()`.

But help from an experienced planner developer is highly appreciated !

Reply:
@qw4990 this is reworked now and the PR is now ready for review! (Now it is called `func (ds *DataSource) AddExtraPhysTblIDColumn() *expression.Column` and is a single function that is called in two different places)